CI failure

CI checks appear to pass on pr creation but merges give errors like 

```
Run npm run test:coverage

> circuit-weather@1.1.1 test:coverage
> vitest run --coverage


 RUN  v1.6.1 /home/runner/work/circuit-weather/circuit-weather
      Coverage enabled with v8

stderr | tests/privacy-modal.test.js > PrivacyModal > Interaction Logic > handles fetch errors gracefully
Failed to load privacy policy: Error: Network error
    at /home/runner/work/circuit-weather/circuit-weather/tests/privacy-modal.test.js:294:44
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:135:14
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:60:26
    at runTest (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:781:17)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runFiles (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:958:5)
    at startTests (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:967:3)
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/vitest/dist/chunks/runtime-runBaseTests.oAvMKtQC.js:116:7

 ✓ tests/privacy-modal.test.js  (33 tests) 71ms
 ✓ tests/circuit-weather-app.test.js  (31 tests) 292ms
 ✓ tests/weather-radar-lifecycle.test.js  (32 tests) 351ms
stderr | tests/weather-radar-lifecycle.test.js > WeatherRadar Lifecycle & Playback > load > always starts polling even if load fails
Radar load failed: Error: Network Error
    at /home/runner/work/circuit-weather/circuit-weather/tests/weather-radar-lifecycle.test.js:302:48
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:135:14
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:60:26
    at runTest (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:781:17)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runFiles (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:958:5)
    at startTests (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:967:3)
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/vitest/dist/chunks/runtime-runBaseTests.oAvMKtQC.js:116:7

stderr | tests/weather-client.test.js > WeatherClient > getForecast > returns error result when fetch fails
Weather fetch failed: Error: Network error
    at /home/runner/work/circuit-weather/circuit-weather/tests/weather-client.test.js:325:45
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:135:14
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:60:26
    at runTest (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:781:17)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runFiles (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:958:5)
    at startTests (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:967:3)
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/vitest/dist/chunks/runtime-runBaseTests.oAvMKtQC.js:116:7

stderr | tests/weather-client.test.js > WeatherClient > getForecast > returns error result when response is not ok
Weather fetch failed: Error: Weather API error
    at WeatherClient.getForecast (/home/runner/work/circuit-weather/circuit-weather/public/src/api/WeatherClient.js:54:41)
    at /home/runner/work/circuit-weather/circuit-weather/tests/weather-client.test.js:337:28
    at runTest (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:781:11)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runFiles (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:958:5)
    at startTests (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:967:3)
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/vitest/dist/chunks/runtime-runBaseTests.oAvMKtQC.js:116:7
    at withEnv (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/vitest/dist/chunks/runtime-runBaseTests.oAvMKtQC.js:83:5)

 ✓ tests/weather-client.test.js  (48 tests) 42ms
stderr | tests/worker.test.js > Worker Logic > Radar API (/api/radar) > handles upstream errors gracefully (empty response)
Upstream Radar API Error: Status 500

stderr | tests/worker.test.js > Worker Logic > Leaflet Assets Proxy (/api/assets/*) > blocks asset with SRI mismatch
SRI Mismatch for leaflet.js: expected 20nQCchB9co0qIjJZRGuk2/Z9VM+kNiyxNV1lvTlZBo=, got AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=

stderr | tests/worker.test.js > Worker Logic > Tile Proxy (/api/tiles/*) > enforces strict upstream content-type (anti-sniffing)
Upstream Tile Invalid Content-Type: text/html (parsed: text/html)

 ✓ tests/worker.test.js  (17 tests) 77ms
stderr | tests/worker-edge-cases.test.js > Worker Edge Cases > /api/track/:id > returns 502 when upstream track fetch fails
Track Fetch Error: Error: Network error
    at /home/runner/work/circuit-weather/circuit-weather/tests/worker-edge-cases.test.js:215:62
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:135:14
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:60:26
    at runTest (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:781:17)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runSuite (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:909:15)
    at runFiles (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:958:5)
    at startTests (file:///home/runner/work/circuit-weather/circuit-weather/node_modules/@vitest/runner/dist/index.js:967:3)
    at file:///home/runner/work/circuit-weather/circuit-weather/node_modules/vitest/dist/chunks/runtime-runBaseTests.oAvMKtQC.js:116:7

 ✓ tests/worker-edge-cases.test.js  (16 tests) 62ms
 ✓ tests/weather-radar-frames.test.js  (16 tests) 276ms
 ✓ tests/sidebar-manager.test.js  (20 tests) 36ms
 ✓ tests/weather-radar-reconcile.test.js  (8 tests) 36ms
 ✓ tests/countdown-timer.test.js  (23 tests) 28ms
 ✓ tests/recentre-control.test.js  (12 tests) 21ms
 ✓ tests/range-circles-logic.test.js  (12 tests) 16ms
stderr | tests/worker-security-content-type.test.js > Worker Security: Strict Content-Type Validation > F1 API Proxy (/api/f1/*) > blocks deceptive type: application/jsonp
Upstream Invalid Content-Type: application/jsonp (parsed: application/jsonp)

stderr | tests/worker-security-content-type.test.js > Worker Security: Strict Content-Type Validation > F1 API Proxy (/api/f1/*) > blocks deceptive type: application/json-evil
Upstream Invalid Content-Type: application/json-evil (parsed: application/json-evil)

stderr | tests/worker-security-content-type.test.js > Worker Security: Strict Content-Type Validation > Radar API Proxy (/api/radar) > blocks deceptive type: application/json-evil
Upstream Radar Invalid Content-Type: application/json-evil (parsed: application/json-evil)

stderr | tests/worker-security-content-type.test.js > Worker Security: Strict Content-Type Validation > Leaflet Assets Proxy (/api/assets/*) > blocks deceptive type: text/javascript-evil
Leaflet Invalid Content-Type (leaflet.js): text/javascript-evil (expected application/javascript or text/javascript)

stderr | tests/worker-security-content-type.test.js > Worker Security: Strict Content-Type Validation > Tile Proxy (/api/tiles/*) > blocks deceptive type: image/png-evil
Upstream Tile Invalid Content-Type: image/png-evil (parsed: image/png-evil)

 ✓ tests/worker-security-content-type.test.js  (8 tests) 61ms
 ✓ tests/weather-radar-polling.test.js  (9 tests) 40ms
 ✓ tests/seo.test.js  (3 tests) 21ms
 ✓ tests/track-layer.test.js  (10 tests) 14ms
 ✓ tests/f1-api.test.js  (8 tests) 20ms
 ✓ tests/session-labels.test.js  (4 tests) 15ms
 ✓ tests/theme-manager.test.js  (8 tests) 23ms
 ✓ tests/map-weather-widget.test.js  (8 tests) 13ms
 ✓ tests/map-manager.test.js  (8 tests) 16ms
 ✓ tests/worker-security.test.js  (17 tests) 12ms
 ✓ tests/worker-utils-helpers.test.js  (12 tests) 42ms
 ✓ tests/status.test.js  (14 tests) 9ms
 ✓ tests/router.test.js  (14 tests) 16ms
 ✓ tests/weather-radar.test.js  (3 tests) 22ms
 ✓ tests/range-circles-geo.test.js  (8 tests) 7ms
 ✓ tests/rate-limiter-expiry.test.js  (2 tests) 12ms
 ✓ tests/utils.test.js  (9 tests) 11ms
 ✓ tests/worker-health-cache.test.js  (1 test) 39ms
 ✓ tests/rate-limiter-leak.test.js  (1 test) 6ms
 ✓ tests/rate-limiter-lru.test.js  (1 test) 7ms
 ✓ tests/rate-limiter-basic.test.js  (4 tests) 6ms
 ✓ tests/rate-limiter-migration-eviction.test.js  (1 test) 4ms

 Test Files  34 passed (34)
      Tests  421 passed (421)
   Start at  23:51:08
   Duration  4.44s (transform 469ms, setup 8ms, collect 1.17s, tests 1.72s, environment 10ms, prepare 4.06s)

 % Coverage report from v8
-------------------|---------|----------|---------|---------|-------------------
File               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
-------------------|---------|----------|---------|---------|-------------------
All files          |   86.63 |    90.28 |   91.61 |   86.63 |                   
 public/src        |   76.41 |    83.13 |      75 |   76.41 |                   
  ...WeatherApp.js |   76.41 |    83.13 |      75 |   76.41 | ...48,636-642,670 
 public/src/api    |     100 |      100 |     100 |     100 |                   
  F1API.js         |     100 |      100 |     100 |     100 |                   
  WeatherClient.js |     100 |      100 |     100 |     100 |                   
 public/src/map    |   88.47 |    91.18 |   89.55 |   88.47 |                   
  MapManager.js    |     100 |      100 |     100 |     100 |                   
  ...therWidget.js |     100 |      100 |     100 |     100 |                   
  RangeCircles.js  |   92.42 |    91.07 |   83.33 |   92.42 | ...49-256,259-263 
  ...treControl.js |     100 |      100 |     100 |     100 |                   
  TrackLayer.js    |   95.48 |    82.75 |     100 |   95.48 | 54-56,80-82       
  WeatherRadar.js  |   83.78 |    90.29 |   86.48 |   83.78 | ...98-800,882-886 
 ...ic/src/routing |     100 |      100 |     100 |     100 |                   
  Router.js        |     100 |      100 |     100 |     100 |                   
 public/src/ui     |   98.73 |      100 |     100 |   98.73 |                   
  ...tdownTimer.js |     100 |      100 |     100 |     100 |                   
  PrivacyModal.js  |   98.85 |      100 |     100 |   98.85 | 18,28             
  ...barManager.js |   97.01 |      100 |     100 |   97.01 | 17-18,25-26       
  ThemeManager.js  |     100 |      100 |     100 |     100 |                   
 public/src/utils  |     100 |      100 |     100 |     100 |                   
  escapeHtml.js    |     100 |      100 |     100 |     100 |                   
  status.js        |     100 |      100 |     100 |     100 |                   
  storage.js       |     100 |      100 |     100 |     100 |                   
 src               |   82.41 |     80.9 |     100 |   82.41 |                   
  worker-utils.js  |   99.28 |    96.77 |     100 |   99.28 | 220-221           
  worker.js        |   77.88 |    73.72 |     100 |   77.88 | ...1018,1031-1039 
-------------------|---------|----------|---------|---------|-------------------

=============================== Coverage summary ===============================
Statements   : 86.63% ( 3857/4452 )
Branches     : 90.28% ( 697/772 )
Functions    : 91.61% ( 142/155 )
Lines        : 86.63% ( 3857/4452 )
================================================================================
ERROR: Coverage for branches (90.28%) does not meet global threshold (90.29%)
Error: Process completed with exit code 1.
```

--- a/public/src/CircuitWeatherApp.js
+++ b/public/src/CircuitWeatherApp.js
@@ -607,7 +607,7 @@ export class CircuitWeatherApp {
     }
 
 
-    renderForecast(weather, sessionTime, sessionId) {
+    renderForecast(weather, sessionTime, sessionId, overrideNow = null) {
         // Updates Sidebar Forecast Panel ONLY
 
         // Guard: If no session is currently selected (e.g., user switched rounds while fetching),
@@ -631,7 +631,7 @@ export class CircuitWeatherApp {
                 const p = unavailable.querySelector('p');
                 if (p) {
                     if (weather.reason === 'too_far' && weather.availableFrom) {
-                        const now = new Date();
+                        const now = overrideNow || new Date();
                         if (weather.availableFrom > now) {
                             const dateStr = weather.availableFrom.toLocaleDateString(undefined, {
                                 month: 'short',

--- a/tests/circuit-weather-app.test.js
+++ b/tests/circuit-weather-app.test.js
@@ -1,25 +1,31 @@
 import { describe, it, expect, vi, beforeEach } from 'vitest';
 
 // --- Global Mocks (same pattern as seo.test.js) ---
-const createMockElement = (id) => ({
-    id,
-    addEventListener: vi.fn(),
-    classList: {
-        add: vi.fn(),
-        remove: vi.fn(),
-        contains: vi.fn().mockReturnValue(true),
-        toggle: vi.fn(),
-    },
-    style: {},
-    setAttribute: vi.fn(),
-    removeAttribute: vi.fn(),
-    textContent: '',
-    value: '',
-    innerHTML: '',
-    disabled: false,
-    querySelector: vi.fn(() => createMockElement('child')),
-    appendChild: vi.fn(),
-});
+const createMockElement = (id) => {
+    const el = {
+        id,
+        addEventListener: vi.fn(),
+        classList: {
+            add: vi.fn(),
+            remove: vi.fn(),
+            contains: vi.fn().mockReturnValue(true),
+            toggle: vi.fn(),
+        },
+        style: {},
+        setAttribute: vi.fn(),
+        removeAttribute: vi.fn(),
+        textContent: '',
+        value: '',
+        innerHTML: '',
+        disabled: false,
+        querySelector: vi.fn((sel) => {
+            if (sel === 'p') return el._p || (el._p = createMockElement('p'));
+            return createMockElement('child');
+        }),
+        appendChild: vi.fn(),
+    };
+    return el;
+};
 
 const documentMock = {
     title: 'Circuit Weather — Live F1 Race Weather Radar & Forecasts',
@@ -474,6 +480,63 @@ describe('CircuitWeatherApp Pure Methods', () => {
 
             // forecastContent should not be modified since session IDs don't match
             expect(app.ui.forecastContent.style.display).toBeUndefined();
+        });
+
+        it('handles "too_far" with availableFrom in the past', () => {
+            const now = new Date('2024-04-02');
+            const weather = {
+                available: false,
+                reason: 'too_far',
+                availableFrom: new Date('2024-04-01')
+            };
+            app.renderForecast(weather, new Date(), 'race', now);
+
+            const p = app.ui.forecastUnavailable.querySelector('p');
+            expect(p.textContent).toBe('Forecast available shortly');
+        });
+
+        it('handles other unavailable reasons', () => {
+            const weather = { available: false, reason: 'unknown' };
+            app.renderForecast(weather, new Date(), 'race');
+
+            const p = app.ui.forecastUnavailable.querySelector('p');
+            expect(p.textContent).toBe('Forecast available closer to session');
+        });
+
+        it('handles "too_far" with availableFrom in the future', () => {
+             const now = new Date('2024-03-31');
+             const weather = {
+                 available: false,
+                 reason: 'too_far',
+                 availableFrom: new Date('2024-04-01T10:00:00')
+             };
+             app.renderForecast(weather, new Date(), 'race', now);
+
+             const p = app.ui.forecastUnavailable.querySelector('p');
+             // .toLocaleDateString output varies by locale, but we expect it to contain the date
+             expect(p.textContent).toContain('Forecast available from');
+        });
+
+        it('finds the closest hourly forecast point', () => {
+            const sessionTime = new Date('2024-03-01T14:00:00Z');
+            const weather = {
+                available: true,
+                current: { temperature_2m: 25 },
+                hourly: [
+                    { time: Math.floor(new Date('2024-03-01T13:00:00Z').getTime() / 1000), temp: 24, windSpeed: 10, windDir: 0, precipProb: 0, code: 0 },
+                    { time: Math.floor(new Date('2024-03-01T14:15:00Z').getTime() / 1000), temp: 26, windSpeed: 10, windDir: 0, precipProb: 0, code: 0 },
+                    { time: Math.floor(new Date('2024-03-01T16:00:00Z').getTime() / 1000), temp: 28, windSpeed: 10, windDir: 0, precipProb: 0, code: 0 },
+                ],
+                units: { temperature_2m: '°C', wind_speed_10m: 'km/h' }
+            };
+
+            app.renderForecast(weather, sessionTime, 'race');
+
+            // The closest point is 14:15 (26 degrees)
+            // We check that the primary temperature value is 26
+            expect(app.ui.forecastContent.innerHTML).toContain('id="weatherTemp">26°C</span>');
+            // 24 should still be in the timeline, but not the primary temp
+            expect(app.ui.forecastContent.innerHTML).toContain('<div>24°</div>');
         });
     });
 

--- a/tests/range-circles-logic.test.js
+++ b/tests/range-circles-logic.test.js
@@ -221,4 +221,41 @@ describe('RangeCircles Logic', () => {
             expect(rangeCircles.rangeColor).toBe('#e10600');
         });
     });
+
+    describe('Visibility & Clearing', () => {
+        beforeEach(() => {
+            rangeCircles = new RangeCircles(mockMap);
+        });
+
+        it('clears all layers including center marker', () => {
+            // Setup layers
+            const mockLabel = { addTo: vi.fn(), remove: vi.fn() };
+            rangeCircles.circles = [mockCircle];
+            rangeCircles.labels = [mockLabel];
+            rangeCircles.centerMarker = mockMarker;
+
+            rangeCircles.clear();
+
+            expect(mockMap.removeLayer).toHaveBeenCalledWith(mockCircle);
+            expect(mockMap.removeLayer).toHaveBeenCalledWith(mockLabel);
+            expect(mockMap.removeLayer).toHaveBeenCalledWith(mockMarker);
+            expect(rangeCircles.circles).toHaveLength(0);
+            expect(rangeCircles.labels).toHaveLength(0);
+            expect(rangeCircles.centerMarker).toBeNull();
+        });
+
+        it('updates visibility by redrawing if center is set', () => {
+            const spy = vi.spyOn(rangeCircles, 'draw').mockImplementation(() => {});
+
+            // No center -> no draw
+            rangeCircles.center = null;
+            rangeCircles.updateVisibility();
+            expect(spy).not.toHaveBeenCalled();
+
+            // Center set -> draw called
+            rangeCircles.center = [0, 0];
+            rangeCircles.updateVisibility();
+            expect(spy).toHaveBeenCalledWith([0, 0]);
+        });
+    });
 });

--- a/tests/track-layer.test.js
+++ b/tests/track-layer.test.js
@@ -40,6 +40,15 @@ vi.stubGlobal('document', documentMock);
 import { TrackLayer } from '../public/src/map/TrackLayer.js';
 import { CIRCUIT_MAP, CONFIG } from '../public/src/config.js';
 
+// Mock config to allow overriding frozen properties
+vi.mock('../public/src/config.js', async (importOriginal) => {
+    const actual = await importOriginal();
+    return {
+        ...actual,
+        CONFIG: { ...actual.CONFIG }
+    };
+});
+
 describe('TrackLayer', () => {
     let trackLayer;
     let mapMock;
@@ -201,5 +210,34 @@ describe('TrackLayer', () => {
         expect(mapMock.removeLayer).toHaveBeenCalledWith(layerMock);
         expect(trackLayer.layer).toBeNull();
         expect(trackLayer.currentCircuitId).toBeNull();
+    });
+
+    it('should handle unknown circuit ID', async () => {
+        const consoleSpy = vi.spyOn(console, 'log').mockImplementation(() => {});
+        await trackLayer.loadTrack('non_existent_circuit');
+        expect(consoleSpy).toHaveBeenCalledWith(expect.stringContaining('No track map found'));
+        expect(fetchMock).not.toHaveBeenCalled();
+    });
+
+    it('should support full URL for trackApi (GitHub mode)', async () => {
+        const originalApi = CONFIG.trackApi;
+        // @ts-ignore - overriding for test
+        CONFIG.trackApi = 'https://raw.githubusercontent.com/bacinger/f1-circuits/master/circuits';
+
+        const circuitId = 'monaco';
+        const geoJsonId = CIRCUIT_MAP[circuitId];
+        fetchMock.mockResolvedValueOnce({
+            ok: true,
+            json: async () => ({ type: 'FeatureCollection', features: [] })
+        });
+
+        await trackLayer.loadTrack(circuitId);
+
+        // Should include .geojson extension in GitHub mode
+        expect(fetchMock).toHaveBeenCalledWith(`${CONFIG.trackApi}/${geoJsonId}.geojson`);
+
+        // Restore
+        // @ts-ignore
+        CONFIG.trackApi = originalApi;
     });
 });

--- a/tests/worker.test.js
+++ b/tests/worker.test.js
@@ -183,6 +183,103 @@ describe('Worker Logic', () => {
       expect(data.radar.past).toEqual([]);
       expect(res.headers.get('X-Upstream-Status')).toBe('500');
     });
+
+    it('handles fetch exceptions gracefully', async () => {
+      mockFetch.mockRejectedValueOnce(new Error('Network error'));
+
+      const req = createRequest('/api/radar');
+      const res = await worker.fetch(req, global.env, global.ctx);
+
+      expect(res.status).toBe(502);
+      const data = await res.json();
+      expect(data.error).toBe('Failed to fetch radar data');
+    });
+
+    it('blocks invalid content-type from upstream radar', async () => {
+      mockFetch.mockResolvedValueOnce(new Response(JSON.stringify({}), {
+        status: 200,
+        headers: { 'Content-Type': 'text/html' }
+      }));
+
+      const req = createRequest('/api/radar');
+      const res = await worker.fetch(req, global.env, global.ctx);
+
+      expect(res.status).toBe(200);
+      const data = await res.json();
+      // Returns empty radar response for invalid content type
+      expect(data.radar.past).toEqual([]);
+    });
+
+    it('sets CORS headers when valid Origin is provided', async () => {
+      mockFetch.mockResolvedValueOnce(new Response(JSON.stringify({ radar: { past: [] } }), {
+        status: 200,
+        headers: { 'Content-Type': 'application/json' }
+      }));
+
+      const req = createRequest('/api/radar', {
+        headers: { 'Origin': 'https://circuit-weather.racing' }
+      });
+      const res = await worker.fetch(req, global.env, global.ctx);
+
+      expect(res.status).toBe(200);
+      expect(res.headers.get('Access-Control-Allow-Origin')).toBe('https://circuit-weather.racing');
+      expect(res.headers.get('Vary')).toBe('Origin');
+    });
+
+    it('handles upstream rate limit (429) for radar', async () => {
+      mockFetch.mockResolvedValueOnce(new Response(JSON.stringify({ error: 'Rate Limit' }), {
+        status: 429,
+        headers: { 'Content-Type': 'application/json', 'Retry-After': '60' }
+      }));
+
+      const req = createRequest('/api/radar');
+      const res = await worker.fetch(req, global.env, global.ctx);
+
+      expect(res.status).toBe(429);
+      const data = await res.json();
+      expect(data.error).toBe('Upstream Rate Limit');
+      expect(res.headers.get('Retry-After')).toBe('60');
+    });
+
+    it('returns cached response for radar with valid origin', async () => {
+        const cachedData = { version: '2.0', host: 'https://x.com', radar: {} };
+        const cacheResponse = new Response(JSON.stringify(cachedData), {
+          headers: { 'Content-Type': 'application/json' }
+        });
+        cacheStore.set('https://api.rainviewer.com/public/weather-maps.json', cacheResponse);
+
+        const req = createRequest('/api/radar', {
+          headers: { 'Origin': 'https://circuit-weather.racing' }
+        });
+        const res = await worker.fetch(req, global.env, global.ctx);
+
+        expect(res.status).toBe(200);
+        expect(await res.json()).toEqual(cachedData);
+        expect(res.headers.get('X-Cache')).toBe('HIT');
+        expect(res.headers.get('Access-Control-Allow-Origin')).toBe('https://circuit-weather.racing');
+        expect(mockFetch).not.toHaveBeenCalled();
+      });
+
+    it('blocks radar request with invalid fetch destination', async () => {
+      const req = new Request('https://circuit-weather.racing/api/radar', {
+        headers: { 'Sec-Fetch-Dest': 'script', 'Sec-Fetch-Site': 'same-origin' }
+      });
+      const res = await worker.fetch(req, global.env, global.ctx);
+      expect(res.status).toBe(403);
+    });
+
+    it('returns radar response without CORS when origin is missing', async () => {
+        mockFetch.mockResolvedValueOnce(new Response(JSON.stringify({ radar: { past: [] } }), {
+          status: 200,
+          headers: { 'Content-Type': 'application/json' }
+        }));
+
+        const req = createRequest('/api/radar'); // No Origin header by default
+        const res = await worker.fetch(req, global.env, global.ctx);
+
+        expect(res.status).toBe(200);
+        expect(res.headers.get('Access-Control-Allow-Origin')).toBeNull();
+      });
   });
 
   describe('Track Proxy (/api/track/*)', () => {

